builder/optimizations.cc

```diff
@@ -454,4 +454,111 @@ class race_condition_fixer : public node_mutator {
 
 stmt fix_buffer_races(const stmt& s) { return race_condition_fixer().mutate(s); }
 
+namespace {
+
+class insert_free_into_allocate : public node_mutator {
+  // Contains the sym of the allocate node + all other buffer nodes which reference it.
+  std::vector<var> names;
+  // If we found some statement which references anything from `names`.
+  bool found = false;
+  // We don't want to insert into the very last statement of the block,
+  // so use this flag to mark that we visited at least one.
+  bool visited_something = false;
+
+public:
+  insert_free_into_allocate(var name) { names.push_back(name); }
+
+  void visit(const block* op) override {
+    // Visit blocks in reverse order.
+    std::vector<stmt> stmts(op->stmts.size());
+    bool changed = false;
+    for (int i = static_cast<int>(op->stmts.size()) - 1; i >= 0; --i) {
+      // Don't mutate the rest if we changed one of the block's statements.
+      if (changed) {
```

Comment:
Should this be `changed || found` ? (Thinking about nested blocks, which can happen if there is an intervening crop or something like that.)

Reply:
You're right, there is no point to mutate if we found something.